app/utils/External.ts

```diff
@@ -195,7 +212,7 @@ export default class External {
         });
     }
 
-    private ajaxCall(
+    private ajaxCall<ResponseBody, RequestBody, RequestQueryParams>(
```

Comment:
I'm not sure what's the purpose for providing unconstrained `RequestBody` and `RequestQueryParams` type parameters. I don't see any benefits. It is possible to call this method with surely invalid parameters, for example `this.ajaxCall('', '', { params: 0 });`.
IMO type for `params` should be enforced by what is expected by `buuildActualUrl` argument.
Similarly, type for `body` should be defined by what's accepted by `JSON.stringify`.

Reply:
OK with `RequestQueryParams`, we can have:
```suggestion
    private ajaxCall<ResponseBody, RequestBody, RequestQueryParams extends QueryParams>(
```
but with `RequestBody` I don't agree it can be anything.

> IMO type for `params` should be enforced by what is expected by `buildActualUrl` argument.

This is the same as `QueryParams`, so I can use this alias of `Record<string, any>` in `buildActualUrl`.

> Similarly, type for body should be defined by what's accepted by JSON.stringify.

And that is `any` 🙂 See:

```
JSON.stringify(value: any, replacer?: ((this:any, key: string, value: any) => any) | undefined, space?: string | number | undefined): string
```

Regarding benefits of providing `RequestBody` - previously `body` was of type `any`, now it can also be `any` if user won't pass second type to this generic function, but if user provides it, you get type safety when providing value for `body` option. And that's the main goal IMO.

Reply:
> OK with `RequestQueryParams`, we can have:
> ...
> Is that what you had in mind?

This would surely help :+1: 

> Regarding benefits of providing `RequestBody` - previously `body` was of type `any`, now it can also be `any` if user won't pass second type to this generic function, but if user provides it, you get type safety when providing value for `body` option. And that's the main goal IMO.

As you pointed out type checking here is optional and depends on the presence of type parameter. Since nothing is enforced on the method caller I can't agree it provides any actual type safety - all it does is enabling certain syntax. It's still a benefit though, so lets leave it :+1:.